Improve aws testing

willthames opened the pull request:
##### SUMMARY
Improve cloud-config-aws.yml and its protection
Improve documentation of setting up IAM user following recent changes

##### ISSUE TYPE
 - Docs Pull Request

##### COMPONENT NAME
testing_integration

##### ANSIBLE VERSION
```
ansible 2.4.0 (devel baa9290d71) last updated 2017/07/17 21:57:27 (GMT +1000)
  config file = None
  configured module search path = [u'/Users/will/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/will/src/ansible/lib/ansible
  executable location = /Users/will/src/ansible/bin/ansible
  python version = 2.7.13 (default, Dec 18 2016, 07:03:39) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]

```

Commit:
Protect AWS credentials from accidental commits

Commit:
Improve documentation for IAM policies

mattclay commented on test/integration/cloud-config-aws.yml.template:
```diff
@@ -14,6 +14,8 @@
 aws_access_key: @ACCESS_KEY
 aws_secret_key: @SECRET_KEY
 security_token: @SECURITY_TOKEN
+aws_region: @REGION
```
Last time I checked, only some of our AWS integration tests support using a region other than us-east-1. That is one of the reasons why ansible-test doesn't support substitution of a region variable. If this is something we're going to support, the tests will need to be fixed as well. Keep in mind that tests will still only run in CI on us-east-1, so there are likely to be regressions in the multi-region support after we fix it.

willthames replied:
All of my tests (so far) seem to be working fine on us-east-2. I'm specifying us-east-2 particularly to identify any obvious flaws assuming us-east-1. That's why I need this extra variable.

dharmabumstead reviewed (approved):
Docs look good to me. Someone from core engineering needs to sign off.

mattclay commented:
CI failure in integration tests due to a yaml syntax error: https://app.shippable.com/github/ansible/ansible/runs/33917/39/console

s-hertel commented on test/integration/cloud-config-aws.yml.template:
```diff
@@ -14,6 +14,8 @@
 aws_access_key: @ACCESS_KEY
 aws_secret_key: @SECRET_KEY
 security_token: @SECURITY_TOKEN
+aws_region: @REGION
```
Since we are using AMIs in tests that require us-east-1 I don't think we can do this without breaking things right now.

s-hertel replied:
Correcting myself a bit:
None of the legacy tests using AMIs have been ported to ansible-test yet. Merging this will make those tests need an alternative to hard coded AMIs before they can be run with CI. It would probably be good to have a solution in mind before merging.

We do have tests with hard coded AZ though:
https://github.com/ansible/ansible/blob/devel/test/integration/targets/ec2_elb_lb/tasks/main.yml#L41

s-hertel replied:
Updated link: https://github.com/ansible/ansible/blob/devel/test/integration/targets/elb_classic_lb/tasks/main.yml#L41

willthames replied:
So in that task `region` is a variable but `zones` are hardcoded? I don't think the problem is with my code here :)

Let me know what you want me to do about this - fixing that test would be the obvious fix for me (although I'll go with `{{ ec2_region }}a` and `{{ ec2_region }}b` as not all regions have four AZs

willthames replied:
I've updated the ELB tests to allow more flexible AZs, although I'm still using three AZs so the tests won't work anywhere with two AZs.

mattclay replied:
If we can update all of our AWS tests to work in multiple regions (at least us-east-1 and us-east-2) that would have some benefits for CI. It would help with rate limiting issues, as we could run Python 2 tests us-east-1 and Python 3 tests in us-east-2.

willthames replied:
@mattclay I'm not aware of any test/integration/targets that now don't work in us-east-2 once this PR is accepted. Is it worth just trying it - happy to fix up any issues, I feel like I'm mostly (if not entirely) there now.

mattclay replied:
I added AWS region support to ansible-test and fixed the .gitignore entries for the temporary cloud config files in a more general way with https://github.com/ansible/ansible/pull/30733.

Your changes to add region support to the tests should work now, although you'll want to rebase first.

Commit:
Update ELB tests to be multi region

Allow AZs to reflect the region in which the tests are run.

This will not work for regions with fewer than three AZs,
but those are relatively rare

willthames commented:
Thanks @mattclay, rebased upon your improvements, should be good to merge when ready.

mattclay commented on test/integration/cloud-config-aws.yml.template:
```diff
@@ -16,6 +16,7 @@ aws_secret_key: @SECRET_KEY
 security_token: @SECURITY_TOKEN
 aws_region: @REGION
 # aliases for backwards compatibility with older integration test playbooks
+ec2_region: '{{ aws_region }}'
```
This is redundant with the last line of the file.

willthames replied:
It is now - you only added that last line on Friday!

willthames replied:
I've removed my commit duplicating your recent work on this :)
